### PR TITLE
Fix shebang and executable paths in gitbash

### LINKED ISSUE #1302: sheband and executable paths incorrect in gitbash
When creating a rez environment with gitbash on Windows, the shebang and executable paths use a single slash which gets escaped and causes the paths to be incorrect.

**Environment**
* OS "windows 10"
* Rez version "2.109.0"
* Rez python version "3.7"

**To Reproduce**
1. Open a gitbash shell on Windows
2. Run a `rez-env`

**Expected behavior**
Taken into a rez environment.

**Actual behavior**
Command not found error

### PR DESCRIPTION
**Pull request by amorphousWaste**
Fixes #1302.

Originally, I thought "shebang and executable paths incorrect in gitbash" was the cause of issues when working with rez in gitbash/Windows, however after investigation and discussion, there were much deeper issues. This PR addresses issues with pathing by adding multiple options for pathing to the needed shells and executor classes.

### CODE REVIEW

**Comment by JeanChristopheMorinPerso** [comment]
@amorphousWaste we will be happy to get the fixes for the linter, but I think it would be more appropriate if it was done in a separate PR.

**Comment by maxnbk** [comment]
linting fixes can be placed against this issue.
https://github.com/AcademySoftwareFoundation/rez/issues/1356

**Comment by amorphousWaste** [comment]
My apologies. I misunderstood the workflow checks and thought that linting was blocking the PR. It has been removed.

**Comment by nerdvegas** [comment]
For visibility please see recent comments: https://github.com/AcademySoftwareFoundation/rez/issues/1302
Would be good to clear this one up, it's been kicking on for a bit.

**Comment by JeanChristopheMorinPerso** [comment]
@amorphousWaste can you update the PR description please to give us an overview of how all this is used, describe what has been tested, potential gray areas, areas not covered, etc please?

That would be greatly appreciated!

**Comment by amorphousWaste** [comment]
This update to the PR adds the functionality of choosing whether or not to modify paths and shell paths via the `rezconfig.py` on a per-shell and per-variable basis. The implementation is delegated to the shell plugin so that how the paths are changed or not is up to the shell. This will apply to environment building and package building. This should conform the previous fixes to fit the points brought up in the various conversations. All previous tests pass in CMD, PowerShell, and Gitbash on Windows. Gitbash has been a struggle to implement and further grey areas are unknown. Hopefully, this will give Gitbash the same functionality as existing shells with no further caveats.

File: src/rez/utils/sourcecode.py
Comment (Jawabiscuit) [inline src/rez/utils/sourcecode.py:108]:
```suggestion
            # letter as per cygpath behavior.
```

File: src/rezplugins/shell/_utils/windows.py
Comment (Jawabiscuit) [inline src/rezplugins/shell/_utils/windows.py:32]:
```suggestion
    # letter as per cygpath behavior.
```

File: src/rezplugins/shell/cmd.py
Comment (Jawabiscuit) [inline src/rezplugins/shell/cmd.py:223]:
```suggestion
        ``is_path`` and ``is_shell_path`` are mutally exclusive.
```

Comment (amorphousWaste) [reply src/rezplugins/shell/cmd.py]: Fixed.

File: src/rezplugins/shell/_utils/windows.py
Comment (Jawabiscuit) [inline src/rezplugins/shell/_utils/windows.py:35]:
```suggestion
            drive_letter_match.expand("/\\1/").lower(), path
```
If emulating cygpath -u, `c:\` or `C:\` should convert to `/c/`

File: src/rez/rezconfig.py
Comment (Jawabiscuit) [inline src/rez/rezconfig.py:536]: A bit of an explanation for the new setting is probably warranted here, no?

File: src/rez/rex.py
Comment (Jawabiscuit) [inline src/rez/rex.py:577]:
```suggestion
        shell_name = cls.name() if hasattr(cls, 'name') else ''
```

File: src/rezplugins/shell/gitbash.py
Comment (Jawabiscuit) [inline src/rezplugins/shell/gitbash.py:116]: Currently, gets called in `VariantBinding` and by the shell's `setenv`. I'm feeling like when a `{root}` path is processed here it should be posix path and in the case of gitbash with the `shell_pathed_env_vars` setting the way it is, this path here should be converted emulating `cypath -m` so `c:\a\b\c` and `C:\a\b\c` convert to `C:/a/b/c`.

File: src/rez/rex.py
Comment (amorphousWaste) [reply src/rez/rex.py]: Fixed in latest commits.

File: src/rezplugins/shell/_utils/windows.py
Comment (amorphousWaste) [reply src/rezplugins/shell/_utils/windows.py]: This is addressed in the latest commits.

File: src/rez/rezconfig.py
Comment (amorphousWaste) [reply src/rez/rezconfig.py:539]: Done.

File: src/rezplugins/shell/gitbash.py
Comment (amorphousWaste) [reply src/rezplugins/shell/gitbash.py:117]: This is addressed in the latest commits.

**Comment by amorphousWaste** [comment]
These updates provide for more flexibility in determining how paths should be reformatted for POSIX, Windows, and Gitbash. This should comply more with how `cygpath` does things while ensuring compatibility with other shells.

**Comment by amorphousWaste** [comment]
Updates the pathing modifications based on testing from @Jawabiscuit .

File: src/rez/rex.py
Comment (Jawabiscuit) [inline src/rez/rex.py:276]: Replace
```
def _add_systemroot_to_env_win32(self, env):
-->     r""" Sets ``%SYSTEMROOT%`` environment variable, if not present
```
with
```
def _add_systemroot_to_env_win32(self, env):
        """ Sets ``%SYSTEMROOT%`` environment variable, if not present
```
in this file.

File: src/rezplugins/shell/sh.py
Comment (Jawabiscuit) [inline src/rezplugins/shell/sh.py:127]: It would probably be good to go ahead and add the missing docstring now.

**Comment by amorphousWaste** [comment]
As per PEP 257:

>  Use r"""raw triple double quotes""" if you use any backslashes in your docstrings. For Unicode docstrings, use u"""Unicode triple-quoted strings""".

**Comment by Jawabiscuit** [comment]
> As per PEP 257:
> 
> > Use r"""raw triple double quotes""" if you use any backslashes in your docstrings. For Unicode docstrings, use u"""Unicode triple-quoted strings""".

TIL: PEP 257! Thanks!

**Comment by JeanChristopheMorinPerso** [comment]
Thanks @amorphousWaste for walking us through your PR during the TSC meeting. Here are the points we would like to see be implemented before we can merge this PR:

* Add a feature flag so that we can have the feature disabled by default. This would be off for a while until we are fully confident that things are working as expected. And it would also help when troubleshooting issues.
* Add logs that will tell us which environment variable were normalized and the paths.
* Add comments in the shell scripts for each normalized paths (or environment variables) affected.
* Add tests.

**Comment by JeanChristopheMorinPerso** [comment]
Hi @amorphousWaste , I see you pushed some new commits to the PR. What's the state of things right now? Can see start reviewing or there is still more work to do before it's ready to review?

**Comment by maxnbk** [comment]
Just want to write that during the last TSC meeting (january) that @amorphousWaste mentioned that the PR is ready for review and testing by interested parties.

@JeanChristopheMorinPerso registered interest in reviewing, and @instinct-vfx registered willingness to try throwing it in front of a few tester individuals in a studio environment, though I imagine that is subject to the whims and priorities of the studio.

**Comment by JeanChristopheMorinPerso** [comment]
@maxnbk I am actively reviewing this PR. I've been doing for a couple of hours now. I should be done today.

**Review by JeanChristopheMorinPerso — CHANGES_REQUESTED** [review]
Thanks for all your work on this @amorphousWaste! I did a first pass on this. There is some things to fix, but the other things are mostly questions, comments, notes, etc.

Note that I haven't tested it yet.

Apart from that, we requested to add some tests. I see some tests were added, but they are pretty light. The path normalization functions are not well covered, corner cases aren't unit tested, etc. I'd also love to see real integration tests, i.e. have packages that we build, resolve in multiple shells and verify that everything looks like it should work works as expected.

Lastly, it's missing documentation. I believe @jasoncscott volunteered to help on that? I would see a new section of our wiki dedicated to that, with example, explanations, etc.

**Comment by JeanChristopheMorinPerso** [comment]
Hey @amorphousWaste @Jawabiscuit , would one of you be willing and have time to pick this up again? If not, please let us know so that we can see if we can find someone else that might be interested.

This PR is important for our community, so we would like to get things moving.

**Comment by Jawabiscuit** [comment]
@JeanChristopheMorinPerso yeah, definitely trying to carve out some time in the near future to revisit this.

I'm wrapping up an extended effort for Linux/Windows Deadline rendering, which also incorporates rez. I'm better positioned now to take another look and continue with the converting to rez on Windows effort soon.

I've enabled gitbash on Windows for myself at least since the last meeting to verify our experience and it has been stable for me working as a developer. Beginning next week I should be able to get more than just me testing and take a look at the comments on this PR in more detail.

I defer to @amorphousWaste on what his plans are, but happy to jump in and take it home if he's preoccupied.

**Comment by JeanChristopheMorinPerso** [comment]
Great news @Jawabiscuit! Just let us know how we can help and of your plans change.

**Comment by predat** [comment]
If I may put my two cents in, i was able to build zlib library with cmake & msvc like i do on linux with a CMakeLists.txt.

File: src/rez/tests/test_shell_utils.py
Comment (jasoncscott) [inline src/rez/tests/test_shell_utils.py:32]: Isn't this the same test as the previous one?  Should it be testing `mixed`?

**Comment by JeanChristopheMorinPerso** [comment]
Closing in favor of #1475.

**Pull request closed without merging**
